maven-core/src/main/java/org/apache/maven/ReactorReader.java

```diff
@@ -324,4 +273,185 @@ private static boolean isTestArtifact(Artifact artifact) {
+                        Files.delete(path);
+                    }
+                }
+                try {
+                    Files.delete(artifactPath);
+                    Files.delete(artifactPath.getParent());
+                    Files.delete(artifactPath.getParent().getParent());
+                } catch (DirectoryNotEmptyException e) {
+                    // ignore
+                }
+            }
+        } catch (IOException e) {
+            LOGGER.error("Error while cleaning project local repository", e);
+        }
+    }
+
+    /**
+     * Retrieve a stream of the project's artifacts
+     */
+    private Stream<Artifact> getProjectArtifacts(MavenProject project) {
+        Stream<org.apache.maven.artifact.Artifact> artifacts = Stream.concat(
+                Stream.concat(
+                        // pom artifact
+                        Stream.of(new ProjectArtifact(project)),
+                        // main project artifact if not a pom
+                        "pom".equals(project.getPackaging()) ? Stream.empty() : Stream.of(project.getArtifact())),
+                // attached artifacts
+                project.getAttachedArtifacts().stream());
+        return artifacts.map(RepositoryUtils::toArtifact);
+    }
+
+    private boolean isRegularFile(Artifact artifact) {
+        return artifact.getFile() != null && artifact.getFile().isFile();
+    }
+
+    private void installIntoProjectLocalRepository(Artifact artifact) {
+        Path target = getArtifactPath(artifact);
+        try {
+            LOGGER.info("Copying {} to project local repository", artifact);
+            Files.createDirectories(target.getParent());
+            Files.copy(
+                    artifact.getFile().toPath(),
+                    target,
+                    StandardCopyOption.REPLACE_EXISTING,
+                    StandardCopyOption.COPY_ATTRIBUTES);
+        } catch (IOException e) {
+            LOGGER.error("Error while copying artifact to project local repository", e);
+        }
+    }
+
+    private Path getArtifactPath(Artifact artifact) {
+        String groupId = artifact.getGroupId();
+        String artifactId = artifact.getArtifactId();
+        String version = artifact.getBaseVersion();
+        String classifier = artifact.getClassifier();
+        String extension = artifact.getExtension();
+        Path repo = getProjectLocalRepo();
+        return repo.resolve(groupId)
+                .resolve(artifactId)
+                .resolve(version)
+                .resolve(artifactId
+                        + "-" + version
+                        + (classifier != null && !classifier.isEmpty() ? "-" + classifier : "")
+                        + "." + extension);
+    }
+
+    private Path getProjectLocalRepo() {
+        Path root = session.getRequest().getMultiModuleProjectDirectory().toPath();
+        return root.resolve("target").resolve("project-local-repo");
```

Comment:
um, this is hardcoded?

Reply:
I'm not really sure this is configurable.  Is it ?

Reply:
https://maven.apache.org/pom.html#Build

Reply:
Default is in super POM https://github.com/apache/maven/blob/20f7c65a520ee4c6a5dde0de4d58e304b1bde442/maven-model-builder/src/main/resources/org/apache/maven/model/pom-4.0.0.xml#L54

Reply:
I could use something like the following to retrieve the build directory of the top most project if it's in the reactor:
```
    private Path getProjectLocalRepo() {
        File mmpd = session.getRequest().getMultiModuleProjectDirectory();
        return session.getProjects().stream()
                .filter(project -> mmpd.equals(project.getBasedir()))
                .findFirst()
                .map(project -> project.getBuild().getDirectory())
                .map(Paths::get)
                .orElseGet(() -> mmpd.toPath().resolve("target"))
                .resolve("project-local-repo");
    }
```

Reply:
Fixed